Support parsing STATUS responses.

comex opened the pull request:
Fixes #185.

My use of ParseError::Unexpected for error messages feels a bit sketchy,
but I wasn't sure how you'd prefer to handle them.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/jonhoo/rust-imap/192)
<!-- Reviewable:end -->

jonhoo commented on src/parse.rs:
```diff
@@ -321,6 +321,69 @@ pub fn parse_mailbox(
     }
 }
 
+pub fn parse_status(
```
We really should write a macro to help write these `parse_*` methods @mordak. They all have basically the same outer structure. Doesn't need to happen as part of this PR though.

jonhoo commented on src/parse.rs:
```diff
@@ -321,6 +321,69 @@ pub fn parse_mailbox(
+) -> Result<Mailbox> {
+    let mut mailbox_opt = None;
+
+    loop {
+        if lines.is_empty() {
+            break mailbox_opt.ok_or_else(|| {
```
Can this just be `while !lines.is_empty()` with the `mailbox_opt.ok_or_else` placed below the loop?

jonhoo commented on src/parse.rs:
```diff
@@ -321,6 +321,69 @@ pub fn parse_mailbox(
+            }
+            Ok((rest, data)) => {
+                lines = rest;
+                if let Some(resp) = try_handle_unilateral(data, unsolicited) {
+                    break Err(resp.into());
```
There is a sort of weird case where we enter this case and `mailbox_opt.is_some()`. I'm not sure if the right thing to do in that instance is to return an `Err` or just ignore the spurious response and give the `Ok`. What do you think?

comex replied:
Well, even if `mailbox_opt` is `None` (i.e. we haven't gotten the expected `STATUS` response yet), that response could still show up later in `lines`.  In particular, if the unknown response is some sort of unsolicited data sent on the server's initiative and unrelated to the `STATUS` command, then the ordering between it and the `STATUS` response would be a matter of chance.

It might be a good idea to make unknown untagged responses non-fatal in all contexts.  The IMAP spec does say: "A client MUST be prepared to accept any server response at all times.  This includes server data that was not requested."  This wording doesn't explicitly differentiate between known response types (that simply don't correspond to requests) and unknown response types, but arguably it could be read to cover both.

But I don't think it makes sense to make them non-fatal only in this case.

jonhoo replied:
I think it's probably okay to return it as an error, because the caller can always then decide what to do next. If they don't want to treat "weird" server messages as errors, they have that option. It only really gets weird in the case where there is _also_ successful return data for the command, since then all the caller can do is call the operation again. But that doesn't work if the operation isn't idempotent.

Honestly not sure what the right thing to do here is, but I agree with you that it's broader than the scope of this PR, so resolving.

cc @mordak if you want to think about what the right thing to do here is too.

jonhoo commented on tests/imap_integration.rs:
```diff
@@ -390,3 +390,12 @@ fn append_with_flags_and_date() {
+fn status() {
+    let mut s = session("readonly-test@localhost");
+    let mb = s.status("INBOX", "(UIDNEXT UIDVALIDITY)").unwrap();
+    assert!(mb.uid_next.is_some());
+    assert!(mb.uid_validity.is_some());
+    assert!(mb.unseen.is_none());
```
Are there more things we can test for here as well that we know will be in the response?

jonhoo commented:
Thanks for taking this up! I left a few notes inline.

comex commented:
I updated the PR:
- Rebased on latest master.
- Addressed PR comments.
- Changed to be more tolerant of unexpected server responses, as the IMAP spec requires:
  - Receiving a STATUS response for the wrong mailbox is no longer an error; it's appended to `unsolicited` like any other unsolicited response.  
  - Receiving multiple STATUS responses for the right mailbox is no longer an error.  It's possible that one of them is the "real" response, and the other ones are unsolicited STATUS responses for the same mailbox, which the server just happened to send around the same time.
  - On the other hand, we can't just stick unsolicited responses with a matching mailbox name into `unsolicited`, because we have no way of knowing which response is the "real" one (if that's even a well-defined concept).  In theory, we could guess that the "real" response is the last response that has (at least) all of the requested attributes, and put other responses into `unsolicited`.  But this would complicate the code, and I doubt it would actually benefit any practical use case.  As a simpler alternative, just merge attributes from all received STATUS responses with a matching mailbox name.  This is the same behavior that `parse_mailbox` already has.

jonhoo reviewed (approved):
I'm happy with this now! Would you mind also opening a PR that adds this change to the changelog? Thanks!
